### DISCUSSION TITLE
我想为 fcitx5 写一个第三方的界面，请问有相关文档吗？

### DISCUSSION THREAD

**Discussion by cathaysia** [fcitx/fcitx5#350] [Q&A] [answered]
起因是我看到有一个插件 https://hub.fastgit.org/rinx/cmp-skkeleton 可以为 nvim 提供一个嵌入式的日文输入法支持，我觉得这个确实非常棒，后来发现也有 vimIm 和  ime.vim 两个中文输入法的插件。vimIm 我不是很清楚，不过 ime.vim，但是它是繁体输入法。因此我想基于 fcitx5 开发一个 vim 中的输入法

我在 fcitx wiki 中发现 fcitx4 似乎是使用 ipc，并且可以借此开发第三方客户端。fcitx5 似乎改成了 DBus。不过请问有一些更详细的文档吗？

**Comment by xingwozhonghua126** [comment]
https://github.com/lilydjwg/fcitx.vim

一般来说界面属于fcitx的ui部分。
感觉和你的需求可能不太一样。

**Comment by xingwozhonghua126** [comment]
https://github.com/fcitx/fcitx5/tree/master/src/ui/classic

可以看看这部分，ui里面的都是界面。但是kimpanel是为了兼容gnome才做的。

**Comment by wengxt** [comment, marked as answer]
ipc 只是表示 inter process communication，后面还是 dbus。你这个实际想做的事情就是创建一个到 fcitx 的 client，你想用 glib 的 api 还是直接调用 dbus 都可以。然后既然界面应该也是 vim 绘制那就还需要设置 ClientSideInputPanel 这个 flag 让界面信息都发回给 vim

**Reply by cathaysia** [reply]
是的。我看到有 DBus，所以猜测 Fcitx5 是一个 CS 模型，但是不知道客户端具体要承担那些任务。只是把字母传给服务端然后从服务端拿到汉字吗？如果这样的话写起来就非常简单。因为 vim 有补全插件，所以我现在是想要先借用补全插件做一个原型，后面体验好的话就不再更改了。

**Reply by wengxt** [reply]
和程序交互的接口是基于按键事件的，不知道补全的接口是什么样的（例如我猜想可能是输入一个字符串，输出一堆字符串），那你就得考虑自己维护一下在字符串变化的时候差值。如果可以有类似按键的接口，那就方便一点，接口提供的东西你可以参考 https://github.com/fcitx/fcitx5-gtk/blob/master/fcitx-gclient/fcitxgclient.h 和 dbus 基本上都对应的

**Reply by wengxt** [reply]
dbus 接口的定义： https://github.com/fcitx/fcitx5-qt/blob/master/qt5/dbusaddons/interfaces/org.fcitx.Fcitx.InputContext1.xml
https://github.com/fcitx/fcitx5-qt/blob/master/qt5/dbusaddons/interfaces/org.fcitx.Fcitx.InputMethod1.xml
首先 CreateInputContext，可以理解为创建了一个会话/连接，然后返回一个 dbus object path，之后你就要继续用这个 object 和 fcitx 互相通信，发送按键，通过dbus signal获取预编辑和候选词列表

**Reply by cathaysia** [reply]
好的，非常感谢。我这部分先研究研究 :)

**Reply by cathaysia** [reply]
请问这两个类的定义是什么呢？我在源代码中没有找到定义
FcitxQtFormattedPreeditList 和 FcitxQtStringKeyValueList

**Reply by tonyfettes** [reply]
@cathaysia 
您好，我也想做一个 neovim 的 fcitx5 输入法 client.

> 借用补全插件做一个原型，后面体验好的话就不再更改了。

我之前写了一个 [Rime 输入法的 coc 插件](https://github.com/tonyfettes/coc-rime)，就是通过提供补全的方式来实现的。但是我认为通过补全实现的输入法有几个问题：
1. 不少输入法框架都会对用户输入进行 debounce，例如 coc.nvim 和 nvim-cmp ，这样输入法（在这里相当于一个 language server）就不能及时地获取用户输入的更新。
2. 补全源的优先级问题。我之前认为中文的输入序列能很好的与英文区分开来，所以将Rime 输入法提供的候选词的优先级设置地较低，这样能够很好的区分开程序中中英文的 token ，但是在撰写文档的时候，一些常用的单字，比如 “的” ，的补全列表里面大部分都会是英文补全项；全角标点符号如果是 trigger character 的话，也会有优先级方面的问题。
3. 在 coc.nvim 中，简单地选取补全项并不会将 textEdit 应用到用户输入上。而我个人的使用习惯是主要使用 <Tab> 进行选取而不是使用 <Enter> 进行确认，所以对于输入法来说基本上每次用户输入都是以取消，而不是以 Commit/Confirm 结束。这可能是 coc.nvim 的特有情况，因为我对其他的补全框架并不是很了解。
4. 使用补全提供候选词列表，那么很可能空格并不能用来确认候选词了。

虽然在这些问题里面有大部分都可以解决，比如可以对汉字和全角标点分别设立补全源和补全优先级，将空格绑定为接受补全项等等，但我认为 debounce 可能还是很难绕过去。

所以，我最近几天尝试写了一个 [原型](https://github.com/tonyfettes/fcitx5.nvim) ，目前仅仅实现了几个最基本的功能，稳定性差，看起来还有比较严重的性能问题，但我个人更加偏好这样接近输入法的体验。

希望能帮到您。

**Reply by cathaysia** [reply]
谢谢，感谢您的答复

**Comment by Ansen88** [comment]
我使用 qt 做了一个带界面的输入法，这个 qt 是通过 dbus 与 fcitx 通讯的，但是这个界面会抢焦点造成输入的结果不能返回到要输入的位置，还请指点

**Comment by QiangF** [comment]
我想给emacs写个fcitx5的client，emacs-fcitx.el，https://github.com/DogLooksGood/emacs-rime/issues/185
我找到launchpad上一个很老的dbus-proxy项目放这里了 https://github.com/QiangF/emacs-dbus-proxy

**Reply by Freed-Wu** [reply]
If you want to realize an [IME inside emacs](https://github.com/rimeinn/ime.nvim#dbus) (embedded fcitx) not an [IME outside emacs](https://github.com/rimeinn/ime.nvim#dbus-1) (system fcitx), you can refer https://github.com/black-desk/fcitx5-ui.nvim/blob/master/lua/fcitx/fcitx.lua#L101-L108


Now, I use https://github.com/rimeinn/rime.nvim , althought two projects share same code https://github.com/rimeinn/rime.nvim/tree/main/packages/ime

**Comment by QiangF** [comment]
@wengxt 有几个比较菜鸟的问题：
1. CreateInputContext 的两个参数有什么意义吗？

     (dbus-call-method :session
                         "org.fcitx.Fcitx5"
                         "/org/freedesktop/portal/inputmethod"
                         "org.fcitx.Fcitx.InputMethod1"
                         "CreateInputContext"
                         '((:struct "Emacs" "efcitx")))

2. ProcessKeyEvent keyval 和 keycode 参数是不是重复了？state, type 分别指什么？有什么用吗？
   ProcessKeyEvent(u keyval, u keycode, u state, b type, u time)
3. 通过dbus signal获取预编辑和候选词列表，是不是通过监听 signal "UpdateClientSideUI"？auxUp, auxDown 分别指什么？

**Reply by wengxt** [reply]
keyval 是经过键盘布局转化之后的
keycode 是键盘扫描码
state 是修饰键状态（ctrl alt等）
type是按下或者松开

auxup和auxdown在界面中预想的展示方式参考

https://fcitx-im.org/wiki/Fcitx_5_Theme#Size_calculation

**Comment by QiangF** [comment]
@wengxt 用下面这段代码，不知道为什么 preedit 是空白？

    ;; 创建 input context，得到 "/org/freedesktop/portal/inputcontext/301"
    (setq fcitx-ic-path (car (dbus-call-method :session
                                               "org.fcitx.Fcitx5"
                                               "/org/freedesktop/portal/inputmethod"
                                               "org.fcitx.Fcitx.InputMethod1" 
                                               "CreateInputContext"
                                               `((:struct "emacs-fcitx" "emacs")))))

    (defvar fcitx-preedit-string nil)
    (defun fcitx-handler-for-client-ui (preedit cursorpos auxUp auxDown candidates
                                                candidateIndex layoutHint hasPrev hasNext)
      (setq fcitx-preedit-string preedit))

    (dbus-register-signal :session
                          "org.fcitx.Fcitx5"
                          fcitx-ic-path
                          ;; "/org/freedesktop/portal/inputcontext/301"
                          "org.fcitx.Fcitx.InputContext1"
                          "UpdateClientSideUI"
                          'fcitx-handler-for-client-ui)

    ;; send key “n”，keysym is 110
    (dbus-call-method :session
                      "org.fcitx.Fcitx5"
                      fcitx-ic-path
                      ;; "/org/freedesktop/portal/inputcontext/301"
                      "org.fcitx.Fcitx.InputContext1"
                      "ProcessKeyEvent"
                      110 0 0 nil (round (time-to-seconds)))

**Reply by wengxt** [reply]
你可能没有设置capability flag

我有一个在tmux里面用fcitx的项目你可以参考一下

https://github.com/wengxt/fcitx5-tmux/blob/master/src/main.cpp

**Reply by QiangF** [reply]
设置 CapabilityFlag::ClientSideInputPanel = (1ULL << 39)，还是不行，虽然成功发生key，但是"UpdateClientSideUI" signal callback 没有被执行，candidates 无法获取，可惜看不懂c++

    (require 'dbus)

    (defvar fcitx-ic-path nil)
    (defvar fcitx-preedit-string "")
    (defvar fcitx-commit-string "")
    (defvar fcitx-candidates nil)

    ;; 1. Create Context
    (setq fcitx-ic-path
          (car (dbus-call-method :session
                                 "org.fcitx.Fcitx5"
                                 "/org/freedesktop/portal/inputmethod"
                                 "org.fcitx.Fcitx.InputMethod1"
                                 "CreateInputContext"
                                 '((:struct "emacs-fcitx" "emacs")))))

    ;; 2. Setup Capabilities (Client Side UI)
    (let* ((Capability_ClientSideControlState (ash 1 2))
           (Capability_ClientSideInputPanel (ash 1 39))
           ;; Combine them. Note: We are EXCLUDING bits related to shared state
           (cap-flag (logior Capability_ClientSideControlState
                             Capability_ClientSideInputPanel)))
      (dbus-call-method :session "org.fcitx.Fcitx5" fcitx-ic-path
                        "org.fcitx.Fcitx.InputContext1" "SetCapability" :uint64
                        cap-flag))


    ;; 3. Register Signal Handler
    (dbus-register-signal :session
                          "org.fcitx.Fcitx5"
                          fcitx-ic-path
                          "org.fcitx.Fcitx.InputContext1"
                          "UpdateClientSideUI"
                          (lambda (preedit cursorpos auxUp auxDown candidates
                                           candidateIndex layoutHint hasPrev hasNext)
                            (setq fcitx-preedit-string preedit)
                            (setq fcitx-candidates candidates)
                            (message "Preedit: %s Candidates: %s" preedit candidates)))

    (dbus-register-signal :session
                          "org.fcitx.Fcitx5"
                          fcitx-ic-path
                          "org.fcitx.Fcitx.InputContext1"
                          "CommitString"
                          (lambda (str)
                            ;; (insert str)
                            ;; Clear preedit since the cycle is finished
                            ;; (setq fcitx-preedit-string "")
                            (setq fcitx-commit-string str)))

    ;; 4. Focus and Test
    (dbus-call-method :session "org.fcitx.Fcitx5" fcitx-ic-path
                      "org.fcitx.Fcitx.InputContext1" "FocusIn")
    (dbus-call-method :session
                        "org.fcitx.Fcitx5"
                        "/controller"
                        "org.fcitx.Fcitx.Controller1"
                        "SetCurrentIM"
                        :string "rime")

    (defun my-fcitx-send-key (keysym)
      "Send a key press to Fcitx."
      (dbus-call-method :session "org.fcitx.Fcitx5" fcitx-ic-path
                        "org.fcitx.Fcitx.InputContext1" "ProcessKeyEvent"
                        :uint32 keysym :uint32 0 :uint32 0 :boolean nil :uint32 0))

    (defun fcitx-send-key (keysym)
      "Sends keysym to Fcitx and returns t if Fcitx handled it."
      (let ((handled (dbus-call-method :session "org.fcitx.Fcitx5" fcitx-ic-path
                                       "org.fcitx.Fcitx.InputContext1" "ProcessKeyEvent"
                                       :uint32 keysym :uint32 0 :uint32 0 :boolean nil :uint32 0)))
        (sleep-for 0.1)
        ;; Fcitx returns a boolean-like integer.
        ;; If handled is non-zero, Fcitx is using it for the IM logic.
        (message "Fcitx handled key %s: %s, candidates %s"
                 keysym
                 (if handled "YES" "NO")
                 fcitx-candidates)))

    ;; Test: Send 'n'
    (fcitx-send-key 110)
    (sleep-for 0.5)
    ;; h
    (fcitx-send-key 104)

**Reply by QiangF** [reply]
貌似根本就没有 updateclientsideui signal

    q@mypc:~$ dbus-monitor "interface='org.fcitx.Fcitx.InputContext1'"
    signal time=1766913197.087688 sender=org.freedesktop.DBus -> destination=:1.70327 serial=4294967295 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
       string ":1.70327"
    signal time=1766913197.087765 sender=org.freedesktop.DBus -> destination=:1.70327 serial=4294967295 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
       string ":1.70327"
    method call time=1766913260.422702 sender=:1.69978 -> destination=org.fcitx.Fcitx5 serial=15 path=/org/freedesktop/portal/inputcontext/44; interface=org.fcitx.Fcitx.InputContext1; member=ProcessKeyEvent
       uint32 104
       uint32 0
       uint32 0
       boolean false
       uint32 0
    signal time=1766913260.422812 sender=:1.5 -> destination=:1.69978 serial=3430 path=/org/freedesktop/portal/inputcontext/44; interface=org.fcitx.Fcitx.InputContext1; member=CurrentIM
       string "Keyboard - English (US)"
       string "keyboard-us"
       string "en"
    signal time=1766913260.534050 sender=:1.5 -> destination=:1.69978 serial=3432 path=/org/freedesktop/portal/inputcontext/44; interface=org.fcitx.Fcitx.InputContext1; member=NotifyFocusOut

用pinyin输入法也一样 https://gist.github.com/QiangF/b722ff6175634dc615d1798599ace9e5

**Reply by QiangF** [reply]
设置了环境变量后不会自动 focusout 了，但是还是无法得到 candidates

    GTK_IM_MODULE="" QT_IM_MODULE="" SDL_IM_MODULE="" INPUT_METHOD="" XMODIFIERS="@im=""" emacs -Q test-fcitx5.el

**Reply by Freed-Wu** [reply]
> 可惜看不懂c++

lua 可能好看懂一点叭。

> 设置capability flag

在这。设置的 

```lua
    capability = capabilities.ClientSideInputPanel + capabilities.KeyEventOrderFix,
```

https://github.com/black-desk/fcitx5-ui.nvim/blob/master/lua/fcitx/fcitx.lua#L28

https://github.com/black-desk/fcitx5-ui.nvim/blob/master/lua/fcitx/fcitx.lua#L105

> 貌似根本就没有 updateclientsideui signal

不可能没有。这是我这边注释了定时器后在 neovim 以外的 lua 中运行的截图：

<img width="1298" height="903" alt="Screenshot_20251228_220242" src="https://github.com/user-attachments/assets/eb974c93-3ec8-4b8b-bebf-0c419ff8a1ec" />

> 设置了环境变量后

我这边跟环境变量一点关系都没有。


不太懂 lisp 。可以先考虑用 lua 或 python 之类的 REPL 调试试试？
我注意的关键是必须要有一个定时器每隔 `fcitx.interval` 时间异步调用一次才能工作。

```lua
lgi.GLib.MainLoop():get_context():iteration()
```

https://github.com/black-desk/fcitx5-ui.nvim/blob/master/lua/fcitx/fcitx.lua#L107

而定时器是由 neovim 自带的 `require'luv'.new_timer()` 提供的。
想提供一个可以在 neovim 以外的 lua 中运行的方案都不好搞定。
更不清楚 emacs 如何提供定时器。

或者你考虑不需要定时器的 rime 也行。 https://github.com/DogLooksGood/emacs-rime
我自己现在也基本上用 http://github.com/rimeinn/rime.nvim 不用这个库了。

**Reply by QiangF** [reply]
就是想整个系统用一个输入法，不想emacs单独弄个rime。很奇怪，即使设置了这个ClientSideInputPanel的capability，还是出现输入法的popup panel，可能是这个原因导致 "UpdateClientSideUI" signal没有触发。可能这个包更接近一些，https://github.com/KawaYww/fcitx5-dbus 但是rust更看不懂了

**Reply by wengxt** [reply]
我觉得你首先应该研究清楚是否 emacs 首先自己已经在调用输入法了，你看见的界面并不一定来自你创建的对应的 client

用 dbus-monitor 观察具体的信息交互吧

**Reply by wengxt** [reply]
> 貌似根本就没有 updateclientsideui signal
> 
> ```
> q@mypc:~$ dbus-monitor "interface='org.fcitx.Fcitx.InputContext1'"
> signal time=1766913197.087688 sender=org.freedesktop.DBus -> destination=:1.70327 serial=4294967295 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameAcquired
>    string ":1.70327"
> signal time=1766913197.087765 sender=org.freedesktop.DBus -> destination=:1.70327 serial=4294967295 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=NameLost
>    string ":1.70327"
> method call time=1766913260.422702 sender=:1.69978 -> destination=org.fcitx.Fcitx5 serial=15 path=/org/freedesktop/portal/inputcontext/44; interface=org.fcitx.Fcitx.InputContext1; member=ProcessKeyEvent
>    uint32 104
>    uint32 0
>    uint32 0
>    boolean false
>    uint32 0
> signal time=1766913260.422812 sender=:1.5 -> destination=:1.69978 serial=3430 path=/org/freedesktop/portal/inputcontext/44; interface=org.fcitx.Fcitx.InputContext1; member=CurrentIM
>    string "Keyboard - English (US)"
>    string "keyboard-us"
>    string "en"
> signal time=1766913260.534050 sender=:1.5 -> destination=:1.69978 serial=3432 path=/org/freedesktop/portal/inputcontext/44; interface=org.fcitx.Fcitx.InputContext1; member=NotifyFocusOut
> ```
> 
> 用pinyin输入法也一样 https://gist.github.com/QiangF/b722ff6175634dc615d1798599ace9e5

这段信息展示的正在使用的输入法是英文啊，哪有输入框呢  ´_>` ……

**Reply by QiangF** [reply]
1. “首先应该研究清楚是否 emacs 首先自己已经在调用输入法了"：
命令前加上空的环境变量可以防止图形界面 emacs 自己调用输入法。
我用的这个命令打开上面的文件：
GTK_IM_MODULE="" QT_IM_MODULE="" SDL_IM_MODULE="" INPUT_METHOD="" XMODIFIERS='@im=""' emacs -Q test-fcitx5.el
test-fcitx5.el 内容如下：
https://gist.github.com/QiangF/d2ee1c5f5c3b11e31bf567b60b8972bb

2. “你看见的界面并不一定来自你创建的对应的 client”：
打开 test-fcitx5.el 后 M-x eval-buffer，我能看到输入框跳出来，显示“你”的候选菜单，可以确定就是这个 client。意味这这几个 capability 设置没用（设置ClientSideControlState,ClientSideInputPanel是不是就不应该有输入框跳出来？）。并且 fcitx-commit-string 也已经变成"你"了，说明有 "CommitString" signal。还可以确认有 "UpdateFormattedPreedit" signal，就是没有"UpdateClientSideUI"。test-fcitx5.el 里把 rime 改成电脑上的中文输入法名称就能运行。

3. 后面这个 dbus-monitor 的输出: https://gist.github.com/QiangF/b722ff6175634dc615d1798599ace9e5
是用 fcitx5 自带的 pinyin 输入法在 Firefox 里输入 nihao 的时候得到的，也没看到 "UpdateClientSideUI" signal。我怀疑是不是 fcitx 还有相关的配置选项没打开。fcitx5 -v 输出：5.1.16，fcitx5-diagnose 输出如下：
https://gist.github.com/QiangF/677375d8b2799837b9a3b8429f05ab26
另外用这个python脚本也没动静 https://gist.github.com/QiangF/27c49ce59c3b83c89dd1cc0ddd84dca8

**Reply by wengxt** [reply]
哦，你的主要问题是没有传递 display ，所以自动加入了 X11 分组，你得手动传递一个匹配不上的，像这样

```
method call time=1767078324.599308 sender=:1.1045 -> destination=org.freedesktop.portal.Fcitx serial=5 path=/org/freedesktop/portal/inputmethod; interface=org.fcitx.Fcitx.InputMethod1; member=CreateInputContext
   array [
      struct {
         string "program"
         string "tmux"
      }
      struct {
         string "display"
         string "tmux"
      }
   ]
```

qdbus6 org.fcitx.Fcitx5 /controller org.fcitx.Fcitx.Controller1.DebugInfo
Group [x11::0] has 12 InputContext(s)
  IC [dd132f5b072841eba7b8eff885378876] program:plasmashell frontend:dbus cap:16000800072 focus:0
  IC [aad329ce49bc48488896ad9514fe5b8e] program:chromium frontend:dbus cap:6000000032 focus:0
  IC [ed45a32861e34c52b8ad0e315b180627] program:Telegram frontend:dbus cap:6000800072 focus:0
  IC [c7cb80cbc955432d98f20bd81aa7f309] program:quasselclient frontend:dbus cap:16000800072 focus:0
  IC [2a0b14df72eb48a89ff74dceb5398795] program:konsole frontend:dbus cap:16000800072 focus:0
  IC [d5c780843c5a45d5a75b1255d94adb42] program:plasmashell frontend:dbus cap:16000800072 focus:0
  IC [81efcfb55b374a7e9c307aa569348f38] program:konsole frontend:dbus cap:16000800072 focus:0
  IC [1ccbf4b736644aef93c4a5793e3042a1] program:konsole frontend:dbus cap:6000820872 focus:1
  IC [6446e592021241aa906a9027e7cf1fad] program:firefox frontend:dbus cap:6000000072 focus:0
  IC [d363bb4a5ac54f64866f58c6af0cd802] program:firefox frontend:dbus cap:6000000032 focus:0
  IC [24ea4dd3021a480d88dc9907e0fd79ac] program:nextcloud frontend:dbus cap:16000800072 focus:0
  IC [c0f8dab6e3e1466fae434759346574e1] program:konsole frontend:dbus cap:6000820872 focus:0
Input Context without group
  IC [2309a6051dba4e21b793714ff05ce0f3] program:tmux frontend:dbus focus:1

让你的 client 出现在最后这个 Input Context without group 就对了

**Reply by QiangF** [reply]
@wengxt 非常感谢，成功了，我把melpa上的imbot改成这个了，git@github.com:QiangF/imbot.git
下一步就是把exwm桌面管理器及emacs的都用这个输入法前端，规避exwm-xim无法在qt5运行的问题。
